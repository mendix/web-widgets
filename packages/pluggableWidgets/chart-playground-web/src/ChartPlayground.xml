<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.web.chartplayground.ChartPlayground" pluginWidget="true" needsEntityContext="true" offlineCapable="true" supportedPlatform="Web" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../node_modules/mendix/custom_widget.xsd">
    <name>Chart playground</name>
    <description>A helper for easy and quick chart customization</description>
    <helpUrl>https://docs.mendix.com/appstore/widgets/ChartPlayground</helpUrl>
    <properties />
</widget>
