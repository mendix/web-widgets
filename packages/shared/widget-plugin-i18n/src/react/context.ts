import { createContext } from "react";
import { PluginI18n } from "../types";

const contextPath = "com.mendix.widget.plugin.i18n.context" as const;

type I18nContext = React.Context<PluginI18n | null>;

const createI18nContext = (): I18nContext => createContext<PluginI18n | null>(null);

declare global {
    interface Window {
        [contextPath]?: ReturnType<typeof createI18nContext>;
    }
}

export function getGlobalI18nPluginContext(): I18nContext {
    return window[contextPath] ?? createI18nContext();
}
