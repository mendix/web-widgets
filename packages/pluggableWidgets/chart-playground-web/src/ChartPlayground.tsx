import { ReactElement, createElement } from "react";
import { Playground } from "./components/Playground";

export function ChartPlayground(): ReactElement | null {
    return <Playground />;
}
