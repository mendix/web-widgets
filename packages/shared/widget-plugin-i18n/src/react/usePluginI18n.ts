import { useState } from "react";
import { I18nController } from "../controllers/I18nController";
import { PluginI18n } from "../types";

export function usePluginI18n(languageCode: string): PluginI18n | null {
    const [plugin, setPlugin] = useState<PluginI18n | null>(() => {
        new I18nController({
            languageCode,
            onPluginChange: plugin => setPlugin(plugin)
        });

        return null;
    });

    return plugin;
}
