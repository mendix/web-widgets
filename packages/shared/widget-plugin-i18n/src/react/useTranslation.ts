import { useContext } from "react";
import { PluginI18n } from "../types";
import { getGlobalI18nPluginContext } from "./context";

export function useTranslation(): PluginI18n | null {
    return useContext(getGlobalI18nPluginContext());
}
