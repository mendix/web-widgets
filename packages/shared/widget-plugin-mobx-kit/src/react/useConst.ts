import { useRef } from "react";

export function useConst<T>(fn: T): T extends (...args: never) => infer R ? R : T {
    return (useRef<T | null>(null).current ??= typeof fn === "function" ? fn() : fn);
}
