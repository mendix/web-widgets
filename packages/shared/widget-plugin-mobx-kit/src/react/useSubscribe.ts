import { IReactionOptions, reaction } from "mobx";
import { useEffect, useState } from "react";

export function useSubscribe<Store extends object, T, FireImmediately extends boolean = false>(
    init: Store | (() => Store),
    selector: (store: Store) => T,
    options?: IReactionOptions<T, FireImmediately>
): [T, Store] {
    const [store] = useState(() => (typeof init === "function" ? init() : init));
    const [value, setValue] = useState(() => selector(store));

    function setup(): () => void {
        return reaction(() => selector(store), setValue, options);
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
    useEffect(setup, []);

    return [value, store];
}
