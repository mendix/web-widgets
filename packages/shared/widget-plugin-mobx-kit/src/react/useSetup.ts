import { useEffect } from "react";
import { useConst } from "./useConst";

export function useSetup<
    T extends {
        setup(): void | (() => void);
    }
>(fn: () => T): T {
    const obj = useConst(fn);
    useEffect(() => obj.setup(), [obj]);
    return obj;
}
