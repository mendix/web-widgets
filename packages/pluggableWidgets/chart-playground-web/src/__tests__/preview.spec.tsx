import { render } from "@testing-library/react";

import { preview } from "../ChartPlayground.editorPreview";

describe("preview", () => {
    it("renders toggle button", () => {
        const { asFragment } = render(preview());
        expect(asFragment()).toMatchSnapshot();
    });
});
