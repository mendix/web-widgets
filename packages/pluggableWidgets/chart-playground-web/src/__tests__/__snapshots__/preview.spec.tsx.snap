// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preview renders toggle button 1`] = `
<DocumentFragment>
  <div
    class="widget-charts-playground"
  >
    <div
      class="widget-charts-playground-toggle"
    >
      <button
        class="mx-button btn"
      >
        Toggle Editor
      </button>
    </div>
  </div>
</DocumentFragment>
`;
